elastic/kibana
# [Fleet] fix upgradeable agents filter

## Summary

Fix https://github.com/elastic/kibana/issues/117329

Fixing bug where `Upgrade available` filter didn't return the expected results. The results included several empty pages. This is due to the API filtering upgradeable after doing a paginated es query.

**To test:**

1. Enroll `fleet-server` `v8.0` or older from local machine and a few agents with docker to have > 5 in total.
2. Open `Fleet Agents `page, select `Rows per page: 5` (Initially the upgradeable agent is on the 2nd page)
3. Click filter `Upgrade available`
4. Verify that one agent is returned without pagination, and on the top it says `Showing 1 agent`

![image](https://user-images.githubusercontent.com/90178898/142887700-490addac-2e7e-4217-b607-ef518aece91b.png)

**Root cause analysis:**

1. Total count is returned from API, which doesn't consider filters.
- Request: `/api/fleet/agents?page=1&perPage=50&showInactive=false&showUpgradeable=true`
- Response: `{"list":[{"id":"0ee72db8-11d5-4a13-9716-b075adf29afb",..."name":"Julias-MacBook-Pro.local"}],"total":7,"totalInactive":0,"page":1,"perPage":50}`

2. On pagination bug, e.g. total 7 agents, page size 5, filtering on `upgrade available`, the API returns 0 results for first page and 1 result in second page only:
- Page 1 request: `/api/fleet/agents?page=1&perPage=5&showInactive=false&showUpgradeable=true`
- Page 1 response: `{"list":[],"total":7,"totalInactive":0,"page":1,"perPage":5}`

- Page 2 request: `/api/fleet/agents?page=2&perPage=5&showInactive=false&showUpgradeable=true`
- Page 2 response: `{"list":[{"id":"0ee72db8-11d5-4a13-9716-b075adf29afb",...}],"total":7,"totalInactive":0,"page":2,"perPage":5}`

I think the root cause is due to `showUpgradeable` being a filter after the es query:
https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/server/services/agents/crud.ts#L140-L148

Proposed fix for `showUpgradeable=true` queries:
Option 1: covers most cases with not too many agents
- query all agents (up to a max size, e.g. 5000)
- filter upgradeable
- return records for requested page

Option 2: cover all cases
- use a cursor to query all agents in batches (so that there is no limit like 5000)
- filter upgradeable
- collect all upgradeable in an array, and take records for requested page
- this gathers potentially many items in memory

I personally prefer Option 1 and fall back to existing logic, supposing having more than 5000 agents is a rare scenario.

Based on telemetry, there is only a few clusters with agents size > 5000:
https://kibana.telemetry.elastic.co/goto/665b6feac1c4ee1b182212bf936741e0
![image](https://user-images.githubusercontent.com/90178898/142885964-b335902c-9948-4bcb-8b71-e090e1bf6b26.png)


### Checklist
